// @codesee/shared-constants/src/color-system.ts
export enum Black {
  // Black Series
  L030 = "#FAFAFB",
  L050 = "#F0F2F6",
  L100 = "#CDD3DF",
  L200 = "#97A0B8",
  L300 = "#535E79",
  L400 = "#273353",
  L500 = "#021443",
  L700 = "#010E34",
  L900 = "#01061E",
  GITHUB = "#181717",
}

export enum Primary {
  // Primary Series (aka Purple!)
  L030 = "#F9F8FD",
  L050 = "#E7E5F2",
  L100 = "#C4BDE0",
  L200 = "#9D92CB",
  L300 = "#7566B6",
  L400 = "#5845A6",
  L500 = "#3A2496",
  L700 = "#361B83",
  L900 = "#2A0D68",
}

export enum Blue {
  // Blue Series
  L030 = "#F7FAFC",
  L050 = "#E9F3FB",
  L100 = "#C7E1F5",
  L200 = "#A2CEEF",
  L300 = "#60ABE4",
  L400 = "#3B98DE",
  L500 = "#1777C1",
  L700 = "#005495",
  L900 = "#003F70",
}

export enum Magenta {
  // Magenta Series
  L050 = "#FBE0EC",
  L100 = "#F6B3CE",
  L200 = "#F080AE",
  L300 = "#EA4D8E",
  L400 = "#E62675",
  L500 = "#E1005D",
  L700 = "#BB0852",
  L900 = "#90003C",
}
export enum Aqua {
  // Aqua Series
  L030 = "#F7FCFC",
  L050 = "#E9F7F6",
  L100 = "#C7F5F2",
  L200 = "#A2E5E3",
  L300 = "#64CECB",
  L400 = "#0FBEBE",
  L500 = "#00A1A1",
  L700 = "#007981",
  L900 = "#005062",
}

export enum Yellow {
  // Yellow Series
  L050 = "#FFFBDC",
  L100 = "#FFEEA0",
  L200 = "#FFD978",
  L300 = "#FFC756",
  L400 = "#FFA722",
}

export enum White {
  L000 = "#FFFFFF",
  TRANSPARENT = "#FFFFFF00",
}

export enum LegendLight {
  LIGHT_GRAY = "#E7E7E7",
  GRAY = "#C2C2C2",
  LIGHT_PURPLE = "#BACFF2",
  JADE_GREEN = "#C0ECD5",
  BEIGE = "#E8DCDE",
  LIGHT_TEAL = "#A5D5E2",
  SEA_GREEN = "#71D297",
  GRASS_GREEN = "#C3F179",
  LIGHT_LAVENDER = "#E1D8FC",
  TEAL = "#51A0B8",
  HUNTER_GREEN = "#4EA46B",
  BUTTER_YELLOW = "#F9E891",
  LAVENDER = "#B39EF8",
  OCEAN_BLUE = "#5687E0",
  CYAN = "#5CD6F1",
  GOLD_YELLOW = "#EDC700",
  ROSE = "#E897B1",
  BROWN = "#A27F4B",
  LIGHT_ROSE = "#F7CAC0",
  SCARLET = "#FF4646",
  HOT_PINK = "#F355A1",
  ORANGE = "#EF7C48",
  YELLOW_GREEN = "#E2E7B6",
  YELLOW_GREEN_DARK = "#93C9B2",
}

export enum LegendDark {
  PURPLE = "#6540DE",
  RUST = "#A24B59",
}

export enum LegendTextColor {
  LIGHT = "#ffffff",
  DARK = "rgba(0, 0, 0, .8)",
}

export enum PRDiagramColor {
  GREEN = "#AAE4B0",
  ORANGE = "#FFCD82",
  RED = "#FF98C3",
  LIGHT_GREEN = "#D2EFD5",
  LIGHT_ORANGE = "#FAE3BE",
  LIGHT_RED = "#FCD1E4",
  DARK_GREEN = "#008243",
  DARK_RED = "#E10079",
}

export enum PRReviewColors {
  BLACK = "#24292f",
  GRAY = "hsl(210deg 18% 87%)",
  BLUE = "#ddf4ff",
  BLUE_WITH_EMPHASIS = "#54aeff66",
  BLUE_DARK = "#0969da",
  GREEN = "#ccffd8",
  GREEN_WITH_EMPHASIS = "#abf2bc",
  RED = "#ffebe9",
  RED_WITH_EMPHASIS = "rgba(255,129,130,0.4)",
}

export type Color =
  | Black
  | Primary
  | Blue
  | Magenta
  | Aqua
  | Yellow
  | White
  | LegendLight
  | LegendDark
  | PRDiagramColor
  | PRReviewColors;

export enum Insight6StepColorScale {
  VALUE_1 = Yellow.L400,
  VALUE_2 = Yellow.L200,
  VALUE_3 = LegendLight.YELLOW_GREEN,
  VALUE_4 = LegendLight.YELLOW_GREEN_DARK,
  VALUE_5 = Blue.L300,
  VALUE_6 = Blue.L500,
}

export enum Insight5StepColorScale {
  VALUE_1 = Yellow.L400,
  VALUE_2 = Yellow.L200,
  VALUE_3 = LegendLight.YELLOW_GREEN,
  VALUE_4 = Blue.L300,
  VALUE_5 = Blue.L500,
}

export enum BoxShadow {
  CARD = "0px 4px 12px rgba(0, 0, 0, 0.1)",
}
