// @codesee/shared-constants/src/index.ts
/*

DDDDDDDDDDDDD                  AAA               NNNNNNNN        NNNNNNNN        GGGGGGGGGGGGGEEEEEEEEEEEEEEEEEEEEEERRRRRRRRRRRRRRRRR
D::::::::::::DDD              A:::A              N:::::::N       N::::::N     GGG::::::::::::GE::::::::::::::::::::ER::::::::::::::::R
D:::::::::::::::DD           A:::::A             N::::::::N      N::::::N   GG:::::::::::::::GE::::::::::::::::::::ER::::::RRRRRR:::::R
DDD:::::DDDDD:::::D         A:::::::A            N:::::::::N     N::::::N  G:::::GGGGGGGG::::GEE::::::EEEEEEEEE::::ERR:::::R     R:::::R
  D:::::D    D:::::D       A:::::::::A           N::::::::::N    N::::::N G:::::G       GGGGGG  E:::::E       EEEEEE  R::::R     R:::::R
  D:::::D     D:::::D     A:::::A:::::A          N:::::::::::N   N::::::NG:::::G                E:::::E               R::::R     R:::::R
  D:::::D     D:::::D    A:::::A A:::::A         N:::::::N::::N  N::::::NG:::::G                E::::::EEEEEEEEEE     R::::RRRRRR:::::R
  D:::::D     D:::::D   A:::::A   A:::::A        N::::::N N::::N N::::::NG:::::G    GGGGGGGGGG  E:::::::::::::::E     R:::::::::::::RR
  D:::::D     D:::::D  A:::::A     A:::::A       N::::::N  N::::N:::::::NG:::::G    G::::::::G  E:::::::::::::::E     R::::RRRRRR:::::R
  D:::::D     D:::::D A:::::AAAAAAAAA:::::A      N::::::N   N:::::::::::NG:::::G    GGGGG::::G  E::::::EEEEEEEEEE     R::::R     R:::::R
  D:::::D     D:::::DA:::::::::::::::::::::A     N::::::N    N::::::::::NG:::::G        G::::G  E:::::E               R::::R     R:::::R
  D:::::D    D:::::DA:::::AAAAAAAAAAAAA:::::A    N::::::N     N:::::::::N G:::::G       G::::G  E:::::E       EEEEEE  R::::R     R:::::R
DDD:::::DDDDD:::::DA:::::A             A:::::A   N::::::N      N::::::::N  G:::::GGGGGGGG::::GEE::::::EEEEEEEE:::::ERR:::::R     R:::::R
D:::::::::::::::DDA:::::A               A:::::A  N::::::N       N:::::::N   GG:::::::::::::::GE::::::::::::::::::::ER::::::R     R:::::R
D::::::::::::DDD A:::::A                 A:::::A N::::::N        N::::::N     GGG::::::GGG:::GE::::::::::::::::::::ER::::::R     R:::::R
DDDDDDDDDDDDD   AAAAAAA                   AAAAAAANNNNNNNN         NNNNNNN        GGGGGG   GGGGEEEEEEEEEEEEEEEEEEEEEERRRRRRRR     RRRRRRR

This package will get published which means none of the data in here is secure. It should *not* be used to store secrets of ANY kind.

 */

///////////////////////////////// LOCAL STORAGE ////////////////////////////////
export const LOCAL_STORAGE_AUTH_MODE_KEY = "CodeSee.authMode";
export const LOCAL_STORAGE_RECORDING_ID_KEY = "CodeSee.recordingId";
export const LOCAL_STORAGE_RECORDING_TIMESTAMP = "CodeSee.recordingTimestamp";
export const LOCAL_STORAGE_WRITE_TOKEN_KEY = "CodeSee.writeToken";
export const LOCAL_STORAGE_VERBOSE_KEY = "CodeSee.verboseData";
export const LOCAL_STORAGE_CONFIG_CACHE_KEY = "CodeSee.recordingConfig";
export const LOCAL_STORAGE_FAB_POSITION_KEY = "CodeSee.fabPosition";

/////////////////////////// TIMOUT CONSTANTS ///////////////////////////////////
export const TRACK_ACROSS_PAGE_LOAD_TIMEOUT = 300_000; // 5 minutes in ms

/////////////////////////// DATA VERBOSITY CONSTANTS ///////////////////////////
// NOT_RECORDING isn't really a data verbosity level, but it's grouped here
// because CodeSeeTrackerImpl.recording toggles between it and the verbosity
// levels. It's important that NOT_RECORDING be false in a boolean context
// vs. the verbosity levels, which must be true.
export const NOT_RECORDING = 0;
export const VERBOSITY_NO_DATA = 100;
export const VERBOSITY_LOW_DATA = 500;
export const VERBOSITY_MAX = 1000;

//////////////////////////////////// AUTH //////////////////////////////////////
export const DEFAULT_USER_EMAIL = "non-user@internal";
export const RTCE_SERVER_USER_ID = "rtce-server";
export const OKTA_EXPECTED_AUDIENCE = "api://default";
export const AUTH_DUMMY_TOKEN_SINGLE_USER = "<CodeSee:SingleUser>";
// When a user connects with SSO, but the identity provider
// does not return an email address, we shim in a fake
// email address with the following suffix.
export const SSO_USER_FAKE_EMAIL_SUFFIX = "@githubtest.com";

export const SCOPES = {
  API_TOKENS: {
    WRITE: "codesee.api_tokens.write",
    MANAGE: "codesee.api_tokens.manage",
  },
  RECORDINGS: {
    READ: "codesee.recordings.read",
    WRITE: "codesee.recordings.write",
    MANAGE: "codesee.recordings.manage",
  },
  STORIES: {
    READ: "codesee.stories.read",
    MANAGE: "codesee.stories.manage",
  },
  ARCH_DIAGRAMS: {
    READ: "codesee.arch_diagrams.read",
    MANAGE: "codesee.arch_diagrams.manage",
    WRITE: "codesee.arch_diagrams.write",
  },
};

export const DEFAULT_SCOPES = [
  // OKTA Scopes
  "openid",
  "profile",
  "email",

  // CodeSee Scopes
  SCOPES.API_TOKENS.WRITE,
  SCOPES.API_TOKENS.MANAGE,
  SCOPES.RECORDINGS.READ,
  SCOPES.RECORDINGS.WRITE,
  SCOPES.RECORDINGS.MANAGE,
  SCOPES.STORIES.READ,
  SCOPES.STORIES.MANAGE,
  SCOPES.ARCH_DIAGRAMS.READ,
  SCOPES.ARCH_DIAGRAMS.MANAGE,
  SCOPES.ARCH_DIAGRAMS.WRITE,
];

/////////////////////// BACKEND/FRONTEND INTEGRATION ///////////////////////////
export const AUTHORIZATION_HEADER_NAME = "X-CodeSee-Authorization";
export const TRACING_HEADER_NAME = "X-CodeSee-Tracing";
export const TRACING_HEADER_NAME_REGEX = /X-CodeSee-Tracing/i;
export const TRACING_HEADER_FORMAT_COMPOSITE_SESSION_ID = "1";
export const TRACING_HEADER_FORMAT_SIMPLE_SESSION_ID = "2";

/////////////////////// ERROR MESSAGES /////////////////////////////////////////
export const MESSAGE_QUEUE_OVERFLOW = "queue overflow";

///////////////////// ERROR CODES ///////////////////
export const SERVER_ERROR_CODES = {
  NO_GITHUB_APP_INSTALLATION: "NoGitHubAppInstallation",
  INVALID_BILLING_ACCOUNT_ID: "InvalidBillingAccountId",
  INSUFFICIENT_GH_PRIVILEGES_FIRST_MAP:
    "InsufficientGitHubPermissionsForFirstMap",
  INSUFFICIENT_GH_PRIVILEGES_SECOND_MAP:
    "InsufficientGitHubPermissionsForSecondMap",
  GITHUB_ACTIONS_NOT_ENABLED: "GitHubActionsNotEnabled",
  GITHUB_REPO_IS_ARCHIVED: "GitHubRepoIsArchived",
};

export const GITHUB_ERROR_CODES = {
  AUTHOR_CANNOT_APPROVE: "AuthorCannotApprove",
  AUTHOR_CANNOT_REQUEST_CHANGES: "AuthorCannotRequestChanges",
  FORBIDDEN: "Forbidden",
  REVIEW_PENDING: "ReviewPending",
  // UNPROCESSABLE_ENTITY is our generic fallback code
  UNPROCESSABLE_ENTITY: "UnprocessableEntity",
} as const;

/**
 * The default value for a GoJS.node.labelKey when it doesn't have a color
 */
export const DEFAULT_GOJS_LABEL_KEY = "0" as const;

export * from "./color-system";
export * from "./fonts";
export * from "./insights";

export const CODESEE_REGISTRATION_URL = "https://login.codesee.io/" as const;
export const CODESEE_RTCE_AUTOMATION_USERNAME = "CodeSee Automated Updater" as const;

/**
 * Users can reach out to this email for support
 */
export const CODESEE_SUPPORT_EMAIL = "support@codesee.io";

/**
 * URLs to the CodeSee docs
 */
export const CODESEE_DOCS_URLS = {
  HOME: "https://docs.codesee.io",
  INSTALLATION: "https://docs.codesee.io/docs/installation",
  GETTING_STARTED: "https://docs.codesee.io/docs/getting-started",
  ONBOARDING_CRA:
    "https://docs.codesee.io/docs/configuring-recordings-with-create-react-app",
  ONBOARDING_TYPESCRIPT:
    "https://docs.codesee.io/docs/typescript-but-without-babel-installation",
  ONBOARDING_NUXT: "https://docs.codesee.io/docs/nuxtjs-installation",
  ONBOARDING_NEXT: "https://docs.codesee.io/docs/nextjs-installation",
  ONBOARDING_STORYBOOK: "https://docs.codesee.io/docs/storybook-installation",
  ONBOARDING_GATSBY: "https://docs.codesee.io/docs/gatsby-installation",
  ONBOARDING_EMBER:
    "https://docs.codesee.io/docs/ember-installation-instructions-experimental",
  ONBOARDING_GENERIC: "https://docs.codesee.io/docs/generic-installation",
  REVIEW_MAPS_GUIDE: "https://docs.codesee.io/docs/user-guide",
  CODEBASE_MAPS_LEARN_MORE: "https://docs.codesee.io/docs/codebase-maps",
  TRIGGERS_LEARN_MORE: "https://docs.codesee.io/docs/triggers-overview",
};

/**
 * URLs to the codesee learn website
 */
export const CODESEE_LEARN_URLS = {
  HOME: "https://learn.codesee.io/",
  EXPLORE:
    "https://learn.codesee.io/types-of-codesee-maps-for-teams/#explore-learn-map-for-personal-onboarding",
  TEACH:
    "https://learn.codesee.io/types-of-codesee-maps-for-teams/#onboarding-map",
  OWNERSHIP:
    "https://learn.codesee.io/types-of-codesee-maps-for-teams/#ownership-map",
  REFACTOR:
    "https://learn.codesee.io/types-of-codesee-maps-for-teams/#technical-debt-map",
};

/**
 * Export our shared Tailwind configuration
 */
export { tailwindConfig } from "./tailwind.config";
