// @codesee/shared-constants/src/tailwind.config.ts
import {
  Aqua,
  Magenta,
  Black,
  Blue,
  Primary,
  Yellow,
  BoxShadow,
} from "./color-system";

export const tailwindConfig = {
  content: [],
  theme: {
    // Only allow our CodeSee colors
    colors: {
      "black-30": Black.L030,
      "black-50": Black.L050,
      "black-100": Black.L100,
      "black-200": Black.L200,
      "black-300": Black.L300,
      "black-400": Black.L400,
      "black-500": Black.L500,
      "black-700": Black.L700,
      "black-900": Black.L900,
      "primary-30": Primary.L030,
      "primary-50": Primary.L050,
      "primary-100": Primary.L100,
      "primary-200": Primary.L200,
      "primary-300": Primary.L300,
      "primary-400": Primary.L400,
      "primary-500": Primary.L500,
      "primary-700": Primary.L700,
      "primary-900": Primary.L900,
      "blue-30": Blue.L030,
      "blue-50": Blue.L050,
      "blue-100": Blue.L100,
      "blue-200": Blue.L200,
      "blue-300": Blue.L300,
      "blue-400": Blue.L400,
      "blue-500": Blue.L500,
      "blue-700": Blue.L700,
      "blue-900": Blue.L900,
      "aqua-30": Aqua.L030,
      "aqua-50": Aqua.L050,
      "aqua-100": Aqua.L100,
      "aqua-200": Aqua.L200,
      "aqua-300": Aqua.L300,
      "aqua-400": Aqua.L400,
      "aqua-500": Aqua.L500,
      "aqua-700": Aqua.L700,
      "aqua-900": Aqua.L900,
      "magenta-50": Magenta.L050,
      "magenta-100": Magenta.L100,
      "magenta-200": Magenta.L200,
      "magenta-300": Magenta.L300,
      "magenta-400": Magenta.L400,
      "magenta-500": Magenta.L500,
      "magenta-700": Magenta.L700,
      "magenta-900": Magenta.L900,
      "yellow-50": Yellow.L050,
      "yellow-100": Yellow.L100,
      "yellow-200": Yellow.L200,
      "yellow-300": Yellow.L300,
      "yellow-400": Yellow.L400,
      transparent: "transparent",
      white: "white",
      current: "currentColor",
    },
    extend: {
      "shadow-md": BoxShadow.CARD,
      borderRadius: {
        "3xl": "1.25rem",
        "4xl": "1.5rem",
      },
      outline: {
        primary: [`2px solid ${Primary.L500}`, "2px"],
      },
    },
  },
  plugins: [],
};
