// codesee/dist/map/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.matchConfig = exports.madgeExtensions = exports.filenamesThatTriggerFullMapSet = exports.fileFormatVersion4 = exports.fileFormatVersion3 = exports.exclusions = void 0;
const madgeExtensions = ["ts", "js", "jsx", "tsx"];
exports.madgeExtensions = madgeExtensions;
const exclusions = [];
exports.exclusions = exclusions;
const filenamesThatTriggerFullMapSet = new Set([
    ".codesee.json",
    "package.json",
    "tsconfig.json",
    "webpack.config.js",
    "webpack.config.json",
]);
exports.filenamesThatTriggerFullMapSet = filenamesThatTriggerFullMapSet;
const fileFormatVersion3 = "3";
exports.fileFormatVersion3 = fileFormatVersion3;
const fileFormatVersion4 = "4";
exports.fileFormatVersion4 = fileFormatVersion4;
const matchConfig = {
    contains: false,
};
exports.matchConfig = matchConfig;
// As of ENG-419, we always generate a version 3 map.
// Version 3 maps have a mode attribute which is either
// "full" or "incremental". A full version 3 map is
// a supersest of the equivalent version 2 map.
// The original `fileFormatVersion` variable is no
// longer use so it's not exported, but we keep it here
// for historical context.
/* eslint-disable @typescript-eslint/no-unused-vars */
// @ts-ignore
const fileFormatVersion = "2";
